### DISCUSSION TITLE
Review format of lyrics in `.mid`

### DISCUSSION THREAD

**Discussion by TheBoxyBear** [TheBoxyBear/charttools#80] [Ideas]
According to the [`.mid` format reference](https://github.com/TheNathannator/GuitarGame_ChartFormats/blob/main/doc/FileFormats/.mid/Standard/Vocals.md), text data for vocals is stored as events that may be shared across tracks. If so, the model should be adjusted accordingly to not allow different lyrics for each track.

**Comment by TheNathannator** [comment]
The lyric text events themselves are not shared across tracks, only the lyric phrase markers are. Each harmony part has its own set of lyrics, including HARM3.

**Comment by TheBoxyBear** [comment]
So global: Start and end of each phrase 

Per standard/harmony track: Positions, pitches and text.

@TheNathannator Know of any midi chart with harmonies for reference?

**Comment by TheNathannator** [comment]
A good amount of the Rock Band 3 setlist has them, along with various RB3 customs.

That list of associations sounds correct off the top of my head, don't think there's anything else to note for that.

**Comment by TheBoxyBear** [comment]
There's also the sharing of SP across harmonies and that seems to be a cascade of phrases, 1 defining for 1/2, 2 for 3 and 3 having no phrases.

Considering cutting vocals altogether for the initial release while a new model is imagined, but such a model also seems difficult to design. It would either be complex to force correct states, or have a lots of runtime error cases when writing. Either way, it would almost definitely not fit in the mold of `Instrument` classes but this isn't a huge issue as vocals can always be treated as a distinct component.

**Comment by TheBoxyBear** [comment]
Found song with hamonies: Amy Winehouse - Rehab

**Comment by TheBoxyBear** [comment]
```mermaid
---
title: Vocals
---
classDiagram

    class PhraseMarker {
        uint Start
        uint End
    }

    class VocalsNote{
        uint Position
        uint Length
        VocalsPitch Pitch
        string Text
    }

    class VocalsTrack {
        List[PhraseMarker] Phrases
    }

    VocalsTrack <|-- StandardVocalsTrack

    class StandardVocalsTrack {
        List[VocalsNote] Notes
    }

    VocalsTrack <|-- HarmonicsVocalsTrack

    class HarmonicsVocalsTrack {
        List[VocalsNote] Harmonics1
        List[VocalsNote] Harmonics2
        List[VocalsNote] Harmonics3
    }

    class Vocals {
        StandardVocalsTrack Standard;
        HarmonicsVocalsTrack Harmonics;
    }
```

**Reply by TheBoxyBear** [reply]
Lists of notes should be replaced with a class as afaik from the docs, each one can have local events. However, a name would need to be found other than `Track` to differentiate the two levels.
